File: 257. Binary Tree Paths/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
var binaryTreePaths = function(root) {
  const res = [];
  const cur = [];
  if (root === null) return res;
  function dfs(node) {
    cur.push(node.val);
    if (node.left === null && node.right === null) {
      res.push(cur.join("->"));
      cur.pop();
      return;
    }
    if (node.left) dfs(node.left);
    if (node.right) dfs(node.right);
    cur.pop();
  }
  dfs(root);
  return res;
};

File: 257. Binary Tree Paths/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        answer_list = list()
        self._binary_tree_paths(root, list(), answer_list)
        return answer_list

    def _is_leaf(self, node: TreeNode):
        return node.left is None and node.right is None

    def _binary_tree_paths(self,
                           node: TreeNode,
                           curr_path: List[int],
                           answer_list: List[str]) -> None:
        if node is not None and self._is_leaf(node):
            curr_path.append(node.val)
            answer_list.append("->".join(list(map(str, curr_path))))
            curr_path.pop()
        elif node is not None:
            curr_path.append(node.val)
            self._binary_tree_paths(node.left, curr_path, answer_list)
            self._binary_tree_paths(node.right, curr_path, answer_list)
            curr_path.pop()
